mirgecom/profiling.py

```diff
@@ -0,0 +1,175 @@
+THE SOFTWARE.
+"""
+
+from meshmode.array_context import PyOpenCLArrayContext
+import pyopencl as cl
+from pytools.py_codegen import PythonCodeGenerator
+import loopy as lp
+import numpy as np
+from dataclasses import dataclass
+
+__doc__ = """
+.. autoclass:: PyOpenCLProfilingArrayContext
+"""
+
+
+@dataclass
+class ProfileEvent:
+    """Class to hold a profile event that (potentially) has not completed yet."""
+    cl_event: cl._cl.Event
+    program: lp.kernel.LoopKernel
+    args_tuple: tuple
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+    """An array context that profiles kernel executions.
+
+    .. automethod:: finish_profile_events
+    .. automethod:: print_profiling_data
+    .. automethod:: call_loopy
+    """
+
+    def __init__(self, queue, allocator=None) -> None:
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            raise RuntimeError("Profiling was not enabled in the command queue."
+                 "Please create the queue with "
+                 "cl.command_queue_properties.PROFILING_ENABLE.")
+
+        self.profile_events = []
+        self.profile_results = {}
+        self.kernel_stats = {}
+
+    def finish_profile_events(self) -> None:
+        # First, wait for event completion
+        if self.profile_events:
+            cl.wait_for_events([t.cl_event for t in self.profile_events])
+
+        for t in self.profile_events:
+            program = t.program
+            (flops, bytes_accessed) = self.kernel_stats[program][t.args_tuple]
+            time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            self.profile_results.setdefault(program, []).append(
+                dict(time=time, flops=flops, bytes_accessed=bytes_accessed))
+
+        self.profile_events = []
+
+    def print_profiling_data(self) -> None:
+        self.finish_profile_events()
+
+        import pytools
+
+        tbl = pytools.Table()
+
+        tbl.add_row(['Function', 'Calls', 'T_min', 'T_avg', 'T_max',
+            'F_min', 'F_avg', 'F_max', 'M_min', 'M_avg', 'M_max', 'BW_avg'])
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v['time'] for v in value]
+            flops = [v['flops'] for v in value]
+            bytes_accessed = [v['bytes_accessed'] for v in value]
+
+            tbl.add_row([key.name, num_values, min(times),
+                int(mean(times)), max(times), min(flops), int(mean(flops)),
+                max(flops), min(bytes_accessed), int(mean(bytes_accessed)),
+                max(bytes_accessed), round(mean(bytes_accessed)/mean(times), 3)])
+
+        if hasattr(pytools.Table, 'github_markdown'):
+            print(tbl.github_markdown())
+        else:
+            print(tbl)
+
+    def call_loopy(self, program, **kwargs) -> dict:
+        program = self.transform_loopy_program(program)
```

Comment:
We had discussed gathering stats on the untransformed program. (Perhaps in addition to the transformed program.) What do you think?

Reply:
I don't remember exactly what our conclusion was but I'm not sure gathering stats on the untransformed program makes too much sense. The performance of the application depends on the performance of the transformed kernels, so I'm not sure if the untransformed kernels add a lot if info.

Reply:
>  The performance of the application depends on the performance of the transformed kernels, 

That's true, but this is about operation counting. And pre-transform kernels:
* often have rectangular loops that are easier to count with primitive (non-barvinok) tools
* have operation counts that are at least not unrelated to the number of net operations performed by the transformed kernel. Gross disagreements between transformed and untransformed could help raise red flags.

Reply:
let's discuss in the meeting today.

Reply:
Per our discussion today, will postpone this for a future PR.